File: Networking/networkclient.cpp
#include "networkclient.h"
#include "networkpacketmanipulator.h"

using namespace Networking;

NetworkClient::NetworkClient(QString _serverIp, quint16 _serverPort)
	: mServerIp(_serverIp),
	mServerPort(_serverPort),
	mSocket(NULL),
	mPacketCounter(0)
{
	mSocket = new QTcpSocket(this);
	mSocket->setObjectName("NetworkClient");
	QObject::connect(mSocket, SIGNAL(connected()), this, SLOT(socketConnected()));
	QObject::connect(mSocket, SIGNAL(disconnected()), this, SLOT(socketDisconnected()));
	QObject::connect(mSocket, SIGNAL(readyRead()), this, SLOT(socketReadyRead()), Qt::QueuedConnection);
	QObject::connect(mSocket, SIGNAL(error(QAbstractSocket::SocketError)), this, SLOT(socketError(QAbstractSocket::SocketError)));
}

NetworkClient::~NetworkClient(void)
{
	if (mSocket)
		mSocket->deleteLater();

	mSocket = nullptr;
}

QString Networking::NetworkClient::serverIp()
{
	return mServerIp;
}

void NetworkClient::setServerIp(QString _serverIp)
{
	mServerIp = _serverIp;
}

quint16 Networking::NetworkClient::serverPort()
{
	return mServerPort;
}

void NetworkClient::setServerPort(quint16 _serverPort)
{
	mServerPort = _serverPort;
}

bool NetworkClient::connectToServer(int _timeOut)
{
	mSocket->connectToHost(QHostAddress(mServerIp), mServerPort);
	return mSocket->waitForConnected(_timeOut);
}

bool NetworkClient::disconnectFromServer(int _timeOut)
{
	if (mSocket->state() == QAbstractSocket::ConnectedState)
	{
		mSocket->disconnectFromHost();
		return mSocket->waitForDisconnected(_timeOut);
	}
	return true;
}

QTcpSocket::SocketState NetworkClient::state()
{
	return mSocket->state();
}

bool NetworkClient::sendPacket(NetworkPacket & _packet)
{
	if (mSocket->state() != QTcpSocket::ConnectedState)
		return false;

	_packet.setId(++mPacketCounter);

	QList<NetworkPacket> nps;
	if (NetworkPacketManipulator::split(_packet, &nps))
	{
		int len = 0;
		for each(NetworkPacket np in nps)
		{
			QByteArray & ba = np.toBuffer();
			len = ba.length();
			if (len != mSocket->write(ba, len))
				return false;
		}

		return true;
	}

	return false;
}

NetworkClientInfo NetworkClient::clientInfo()
{
	return NetworkClientInfo(mSocket->peerAddress(), mSocket->peerPort());
}

quint32 NetworkClient::packetCounter()
{
	return mPacketCounter;
}

QString NetworkClient::lastError()
{
	return mSocket->errorString();
}

#pragma region slots
void NetworkClient::socketConnected()
{
	emit connected();
}

void NetworkClient::socketDisconnected()
{
	emit disconnected();
}

void NetworkClient::socketReadyRead()
{
	int totalBytes = mSocket->bytesAvailable();

	while (totalBytes >= NetworkingGlobals::FixedBufferSize)
	{
		QByteArray & data = mSocket->read(NetworkingGlobals::FixedBufferSize);
		NetworkPacket np(data);

		if (np.packetType() == NPT_DATA)
		{
			// partial data end check
			if (mPartialDataBuffer.size() > 0)
			{
				mPartialDataBuffer.clear();
				emit error(QAbstractSocket::NetworkError);
			}

			emit packetReceived(np);
		}
		else if (np.packetType() == NPT_PARTIAL_DATA)
		{
			mPartialDataBuffer.append(np);
		}
		else if (np.packetType() == NPT_PARTIAL_DATA_END)
		{
			mPartialDataBuffer.append(np);

			NetworkPacket npj(NPT_DATA);
			if (NetworkPacketManipulator::join(mPartialDataBuffer, &npj))
				emit packetReceived(npj);
			else
				emit error(QAbstractSocket::NetworkError);

			mPartialDataBuffer.clear();
		}

		totalBytes -= data.size();
	}
}

void NetworkClient::socketError(QAbstractSocket::SocketError _socketError)
{
	emit error(_socketError);
}
#pragma endregion ~slots

File: NetworkingDotNet/NetworkClient.cs
/*
TCP Networking Library (.Net Version)

Author: Recep ALTIN 
*/

using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Threading;

namespace NetworkingDotNet
{
    public class NetworkClient
    {
        #region Properties
        public string ServerIp { get; set; }
        public uint ServerPort { get; set; }
        public string LastError { get; protected set; }
        public bool IsConnected { get => (mClient != null ? mClient.Connected : false); }
        #endregion

        #region Events
        public event ClientHandler Disconnected;
        public event ClientPacketReceiveHandler PacketReceived;
        public event ClientErrorHandler Error;
        #endregion

        #region Fields
        protected TcpClient mClient = null;
        protected Thread mThreadListener = null;
        protected uint mPacketCounter = 0;
        protected List<NetworkPacket> mPartialDataBuffer = new List<NetworkPacket>();
        #endregion

        public NetworkClient()
        {
            ServerIp = NetworkingDotNetGlobals.DefaultServerIP;
            ServerPort = NetworkingDotNetGlobals.DefaultServerPort;
        }

        public NetworkClient(string _serverIp, uint _serverPort)
        {
            ServerIp = _serverIp;
            ServerPort = _serverPort;
        }

        public bool ConnectToServer(int _timeOut)
        {
            bool res = false;
            try
            {
                mClient = new TcpClient();
                mClient.Connect(ServerIp, (int)ServerPort);

                if (res = mClient.Connected)
                {
                    mThreadListener = new Thread(new ThreadStart(ListenTCP))
                    {
                        IsBackground = true
                    };
                    mThreadListener.Start();
                }
            }
            catch (Exception ex)
            {
                LastError = ex.Message;
                Error?.Invoke(LastError);
            }

            if (!res)
            {
                DisconnectFromServer(_timeOut);
            }

            return res;
        }

        public bool DisconnectFromServer(int _timeOut)
        {
            bool res = false;

            try
            {
                if (mClient != null)
                {
                    if (mThreadListener != null)
                    {
                        mThreadListener.Interrupt();
                        mThreadListener.Join();
                        mThreadListener = null;
                    }

                    mClient.Close();
                    mClient = null;
                }

                res = true;
            }
            catch (Exception ex)
            {
                LastError = ex.Message;
                Error?.Invoke(LastError);
            }

            return res;
        }

        private void ListenTCP()
        {
            try
            {
                NetworkStream ns = mClient.GetStream();
                while (mClient.Connected)
                {
                    if (ns.DataAvailable)
                    {
                        try
                        {
                            byte[] buffer = new byte[NetworkingDotNetGlobals.FixedBufferSize];
                            ns.Read(buffer, 0, buffer.Length);

                            NetworkPacket np = new NetworkPacket(NetworkPacketType.NPT_NULL);
                            np.FromBuffer(new List<byte>(buffer));

                            if (np.PacketType == NetworkPacketType.NPT_DATA)
                            {
                                // partial data end check
                                if (mPartialDataBuffer.Count > 0)
                                {
                                    uint id = mPartialDataBuffer[0].Id;
                                    mPartialDataBuffer.Clear();
                                    LastError = String.Format("Packet ID:{0} lost.", id);
                                    Error?.Invoke(LastError);
                                }

                                PacketReceived?.Invoke(np);
                            }
                            else if (np.PacketType == NetworkPacketType.NPT_PARTIAL_DATA)
                            {
                                mPartialDataBuffer.Add(np);

                                PacketReceived?.Invoke(np);
                            }
                            else if (np.PacketType == NetworkPacketType.NPT_PARTIAL_DATA_END)
                            {
                                mPartialDataBuffer.Add(np);

                                NetworkPacket npj = new NetworkPacket(NetworkPacketType.NPT_DATA);
                                if (NetworkPacketManipulator.Join(mPartialDataBuffer, ref npj))
                                {
                                    PacketReceived?.Invoke(npj);
                                }
                                else
                                {
                                    LastError = String.Format("Packet ID:{0} corrupt data.", npj.Id);
                                    Error?.Invoke(LastError);
                                }

                                mPartialDataBuffer.Clear();
                            }
                        }
                        catch (SocketException ex)
                        {
                            LastError = ex.Message;
                            Error?.Invoke(LastError);
                            //break;
                        }
                        catch (Exception ex)
                        {
                            LastError = ex.Message;
                            Error?.Invoke(LastError);
                            //break;
                        }
                    }
                }
            }
            catch (ThreadInterruptedException)
            { }

            Disconnected?.Invoke();
        }

        public bool SendPacket(NetworkPacket _packet)
        {
            if(!mClient.Connected)
		        return false;
	
	        _packet.Id = ++mPacketCounter;
	
	        List<NetworkPacket> nps = new List<NetworkPacket>();
            if (NetworkPacketManipulator.Split(_packet, ref nps))
            {
                try
                {
                    NetworkStream ns = mClient.GetStream();
                    foreach (NetworkPacket np in nps)
                    {
                        List<byte> ba = np.ToBuffer();
                        ns.Write(ba.ToArray(), 0, ba.Count);
                        ns.Flush();
                    }
                    return true;
                }
                catch (Exception ex)
                {
                    LastError = ex.Message;
                    Error?.Invoke(LastError);
                }
            }

	        return false;
        }

        public NetworkClientInfo GetInfo()
        {
            IPEndPoint ep = (IPEndPoint)mClient.Client.RemoteEndPoint;
            return new NetworkClientInfo(ep.Address.ToString(), (uint)ep.Port);
        }

        public uint GetPacketCounter()
        {
            return mPacketCounter;
        }

        public override string ToString()
        {
            return string.Format("<NetworkClient ServerIp:{0} ServerPort:{1} IsConnected:{2}>", ServerIp, ServerPort, mClient != null ? mClient.Connected : false);
        }
    }
}
